pkg/statser/statser_internal.go

```diff
@@ -0,0 +1,151 @@
+package statser
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/atlassian/gostatsd"
+)
+
+// InternalHandler is an interface to dispatch metrics to.  Exists
+// to break circular dependencies.
+type InternalHandler interface {
+	DispatchMetric(ctx context.Context, m *gostatsd.Metric) error
+}
+
+// InternalStatser is a Statser which sends metrics to a handler on a best
+// effort basis.  If all buffers are full, metrics will be dropped.  Dropped
+// metrics will be accumulated and emitted as a gauge (not counter).  Metrics
+// sent after the context is closed will be counted as dropped, but never
+// surfaced because it has nowhere to submit them.
+//
+// There is an assumption (but not enforcement) that InternalStatser is a
+// singleton, and therefore there is no namespacing/tags on the dropped metrics.
+type InternalStatser struct {
+	ctx    context.Context
+	wg     *sync.WaitGroup
+	buffer chan *gostatsd.Metric
+
+	tags      gostatsd.Tags
+	namespace string
+	hostname  string
+	handler   InternalHandler
+	dropped   uint64
+}
+
+// NewInternalStatser creates a new Statser which sends metrics to the
+// supplied InternalHandler.  The WaitGroup must be waited on after
+// the context is closed.
+func NewInternalStatser(ctx context.Context, wg *sync.WaitGroup, bufferSize int, tags gostatsd.Tags, namespace, hostname string, handler InternalHandler) Statser {
+	is := &InternalStatser{
+		ctx:       ctx,
+		wg:        wg,
+		buffer:    make(chan *gostatsd.Metric, bufferSize),
+		tags:      tags,
+		namespace: namespace,
+		hostname:  hostname,
+		handler:   handler,
+	}
+	wg.Add(1)
+	go is.run()
```

Comment:
Good catch - had that for when I was creating a new InternalStatser, before the TaggingStatser.